### ISSUE TITLE
Feature request: SSL

### ISSUE THREAD

**Issue by ghazel** [libuv/libuv#100]
Lots people seem to want it, but libuv lacks SSL support. Support for SSL in libuv (or as an add-on to libuv) would make adopting libuv easier. OpenSSL is hard to get right, and thus a lot of effort for the application.

Bonus points for Apple's Security framework, where available.

**Comment by saghul** [comment]
I don't think we want to go this route. Applications should implement it the way they see fit. I know it's hard to get right, but it's impossible to please everyone: why OpenSSL and not gnuTLS / LibreSSL / PolarSSL / what have you? Do we add wrappers for all cert management functions or let people just use OpenSSL's API? What about DTLS? And TLS over Unix domain sockets?

I'm very -1 on this.

/cc @bnoordhuis @piscisaureus

**Comment by ghazel** [comment]
The approach libevent took seems to work fairly well for people. In other languages, Twitsted and EventMachine did a good job also.

I'm not advocating for OpenSSL in particular, but some sort of out-of-the-box TLS solution would be very handy.

**Comment by bnoordhuis** [comment]
SSL/TLS support comes up often; not so much as a feature request but as a 'how do I integrate libuv with openssl?' kind of question on SO and other places.  As such, I think it would be useful to have out-of-the-box support but it doesn't necessarily have to be part of libuv itself; it should be doable as a separate library on top of libuv that lives under the libuv project umbrella but with possibly a different set of maintainers.

If anyone is interested in taking that on, now would be a good time to step up. :-)

> why OpenSSL and not gnuTLS / LibreSSL / PolarSSL / what have you?

curl supports those by means of an abstraction layer that shims the largest common denominator, with some library-specific escape hatches.  Presumably our hypothetical libuv-ssl would employ a similar tactic.

> TLS over Unix domain sockets?

Do people actually do that?  How does verification work?

**Comment by saghul** [comment]
> Do people actually do that? How does verification work?

I have no idea, I made it up on the spot :-)

**Comment by deleisha** [comment]
+1 for considering this as enhancement. I am also assuming DTLS for udp will also included?

**Comment by trevnorris** [comment]
@bnoordhuis So think a better way would be to have an addition that statically compiles into libuv with that extra support? I mean, is there really anything that couldn't be done only using libuv's public API?

**Comment by bnoordhuis** [comment]
@trevnorris Nothing like that, it would be completely unbundled; it'd just be a known-good library that is vetted and maintained in lockstep with libuv proper.

**Comment by saghul** [comment]
A libuv user announced this on the mailing list today: https://github.com/deleisha/libuv-ssl

**Comment by guymguym** [comment]
+1 including DTLS.

Some thoughts:
* Applications *may* want to choose SSL implementation, but I think they will be happy with any quality solution that you can provide.
* What applications definitely do not want do is the entire integration from scratch every time they want to turn on security.
* Security is already a commodity that ships inside every OS, so infrastructures that don't provide it in a simple switch only add friction to its users.

Cheers.

**Comment by yorkie** [comment]
Just announce though that a library supports SSL with libuv: https://github.com/clibs/net :)

**Comment by allspace** [comment]
TLS is important for libuv.

**Comment by deleisha** [comment]
Was planning to work on it once issue #275 is in place.

**Comment by liudf0716** [comment]
TLS support is very important

**Comment by saghul** [comment]
I move to close this.  The plan is to land TLS support in libuv-extras, when ready, not here.  Thoughts @libuv/collaborators ?

**Comment by bnoordhuis** [comment]
I agree.  I'll close out the issue.

**Comment by zhaojh329** [comment]
TLS support is very important

**Comment by yorkie** [comment]
@saghul I would be happy that the TTS part of https://github.com/clibs/net could you merged into libuv-extras if that make sense :)
